Test botan on Travis

wilzbach opened the pull request:
Analog to https://github.com/dlang/dub-registry/pull/214 and https://github.com/dlang/dub/pull/1138

The main idea is to use daily crons and avoid being broken in the future.

ZombineDev commented on travis-ci.sh:
```diff
@@ -2,6 +2,12 @@
+    grep -qF "Have_botan" stream/dub.sdl || echo 'versions "Have_botan"' >> stream/dub.sdl
+    # Make botan mandatory (i.e. non-optional)
+    sed -E 's/(dependency "botan".*) optional=true/\1/' -i stream/dub.sdl
+fi
```
Wow, that's quite an ugly hack. I guess you had no other choice..

wilzbach replied:
Yup, at least I couldn't think of one and this one is not too bad: it's only two lines long and even idempotent.

If wanted I could also add an else clause that idempotently resets this change, but imho (1) this script is only used for the CI anyways and (2) if used locally `git checkout` already works well.

MartinNowak replied:
The Have_botan part shouldn't be necessary if botan is a dependency.
Usually you have to add it to your dub.selections.json to select an optional dependency, so `dub upgrade` and some `jq` command might work, not much better than your sed line though. We definitely need a dub command for that stuff.
Would also make sense to change openssl's `default=true` to `default=false`.
But weirdly enough botan itself lists openssl as dependency, so you can't really get rid of it.
https://github.com/etcimon/botan/blob/32f01e942dc44b8fbbc5eb7fbd285619c57a42af/dub.json#L127

wilzbach replied:
> But weirdly enough botan itself lists openssl as dependency, so you can't really get rid of it.

But only for the `full_openssl` configuration, not by default.

ZombineDev commented on .travis.yml:
```diff
@@ -19,29 +19,28 @@ d:
-  - dmd-2.074.0-rc1
+  - dmd-2.074.0
   - dmd-2.070.2
-  - ldc-1.1.1
```
Can you keep LDC 1.1.1?

wilzbach replied:
The [build matrix](https://travis-ci.org/rejectedsoftware/vibe.d/builds/237027475) is already extremely long. Do you have any specific reason to keep ldc-1.1.1?

wilzbach commented:
Well I think [auto-cancellation of outdated commits](https://blog.travis-ci.com/2017-03-22-introducing-auto-cancellation) should be enabled for this repo:

![image](https://cloud.githubusercontent.com/assets/4370550/26551478/108b3fe2-4483-11e7-9258-0c5b9ebf3810.png)

Moreover as it now takes a very long time per job, I guess we have to be very careful with how many builds on OSX we enable..

Last, but not least, my main point was to show that non-hardcoded CI tags + daily rebuilds are a good way to catch regressions early on as currently Vibe.d doesn't build with `dmd-nightly`, e.g.

```
stream/vibe/stream/memory.d(19,80): Error: cannot implicitly convert expression processAllocator() of type shared(ISharedAllocator) to std.experimental.allocator.IAllocator
stream/vibe/stream/memory.d(48,42): Error: cannot implicitly convert expression processAllocator() of type shared(ISharedAllocator) to std.experimental.allocator.IAllocator
utils/vibe/utils/array.d(715,35): Error: function vibe.utils.array.ArraySet!(Task).ArraySet.AW (IAllocator a) is not callable using argument types (shared(ISharedAllocator))
core/vibe/core/drivers/libevent2.d(656,23): Error: template instance vibe.utils.hashmap.HashMap!(Thread, ThreadSlot, DefaultHashMapTraits!(Thread)) error instantiating
utils/vibe/utils/array.d(715,35): Error: function vibe.utils.array.ArraySet!ulong.ArraySet.AW (IAllocator a) is not callable using argument types (shared(ISharedAllocator))
dmd failed with exit code 1.
```

The first error originates from dlang/phobos#5355
If anyone wants to look at it, on the journey I already discovered a [nasty ICE](https://issues.dlang.org/show_bug.cgi?id=17451).

But there's more fun as LDC >= 1.0.0 segfaults with `libasync`:

```
CoreTaskFiber was terminated unexpectedly: Range violation
```

wilzbach commented:
Removed `dmd-nightly`, `dmd-beta` and `ldc-beta` form the list to avoid breaking Vibe.d by accident. And as Vibe.d is active enough, I removed `dmd` and `ldc` as well ...

s-ludwig commented:
Good news is that the Botan hack can also be replaced by `--override-config vibe-d:tls/botan` now due to #1823.

One thing that still bothers me is that the list of tests is now even longer. What about making this an `includes:` configuration that is just tested with "dmd" and maybe one with "ldc"?

wilzbach commented:
> Good news is that the Botan hack can also be replaced by --override-config vibe-d:tls/botan now due to #1823.

Hmm - how do we trick DUB in ignoring this warning then:

> Package 'vibe-d:tls', marked for configuration override, is not present in dependency graph.

Maybe a workaround isn't that bad after all?

> One thing that still bothers me is that the list of tests is now even longer. What about making this an includes: configuration that is just tested with "dmd" and maybe one with "ldc"?

Makes sense.

s-ludwig commented:
> Package 'vibe-d:tls', marked for configuration override, is not present in dependency graph.

For the unit tests it would just be a matter of specifying it for only those sub packages that use `:tls` directly or indirectly. For the high level tests... it would make sense to run only those tests that use TLS anyway, so we'd probably need a marker file there.

> Maybe a workaround isn't that bad after all?

The other problem is that it doesn't work at all anymore, OpenSSL will still be used as the TLS provider. It could of course be adjusted to simply remove the "openssl" configuration of "tls/dub.sdl".

Geod24 commented:
Closing as Travis is gone.
